Quartz/Engine/Include/Quartz/Utilities/Commander.hpp

```diff
@@ -0,0 +1,143 @@
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#pragma once
+
+#include <array>
+#include <functional>
+#include <string>
+
+namespace qz
+{
+	namespace utils
+	{
+		constexpr std::size_t MAX_COMMANDS_NUMBER  = 100;
+		constexpr std::size_t MAX_ARGUMENTS_NUMBER = 10;
+		typedef std::function<int(
+		    std::array<std::string, MAX_ARGUMENTS_NUMBER> args)>
+		    function;
+
+		/**
+		 * @brief The command book stores commands and information on them to be
+		 * used by a commander
+		 */
+
+		struct CommandBook
+		{
+			std::array<std::string, MAX_COMMANDS_NUMBER> m_command;
+			std::array<std::string, MAX_COMMANDS_NUMBER> m_help;
+			std::array<std::string, MAX_COMMANDS_NUMBER> m_permission;
+			std::array<function, MAX_COMMANDS_NUMBER>    m_functions;
+
+			/**
+			 * @brief Registers a command in the command registry
+			 *
+			 * @param command The keyword for calling the command
+			 * @param permission What permission is required to run this command
+			 */
+			void add(const std::string& command, const std::string& help,
+			         const std::string& permission, function f);
+
+			/**
+			 * @brief Searches for a command
+			 *
+			 * @param command The command to search for
+			 * @return index command is located at or -1 if its not found
+			 */
+			int find(const std::string& command);
+
+			/**
+			 * @brief Gets next open space in book
+			 */
+			int getPage();
+
+		private:
+			/**
+			 * @brief Counter that keep track of next space in the arrays (eg
+			 * next page in the book)
+			 */
+			int m_page;
+		};
+
+		/**
+		 * @brief The Commander handles comand line functions. It accepts an in
+		 * and out stream on creation and can interact with the command line or
+		 * another interface.
+		 *
+		 */
+		class Commander
+		{
+		private:
+			CommandBook   m_book;
+			std::ostream& m_out;
+			std::istream& m_in;
+
+		public:
+			/**
+			 * @brief Initializes a commander that can run and execute commands
+			 *
+			 * @param book The command book that the commander uses to route
+			 * commands to functions
+			 * @param out An output stream the commander outputs results to
+			 * @param in An input stream the commander accepts input to
+			 */
+			Commander(CommandBook& book, std::ostream& out, std::istream& in);
+			~Commander();
+
+			/**
+			 * @brief Calls a command
+			 *
+			 * @param command The keyword for calling the command.
+			 * @return Returns return from function or -1 if function not found
+			 */
+			int run(const std::string& command,
+			        const std::array<std::string, MAX_ARGUMENTS_NUMBER>&& args);
+
+			/**
+			 * @brief Returns helpstring for command
+			 *
+			 * @param args array of input, args[0] is the command helpstring is
+			 * returned for
+			 * @return Returns 1 if successful and -1 if not (most often when
+			 * command is not found)
+			 */
+			int help(const std::array<std::string,
```

Comment:
Would it make more sense to return a boolean (or even an enum here)? It seems that the only two states are 1 (OK) and - (Not OK)